src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Simple.cs

```diff
@@ -0,0 +1,573 @@
+            byte[] received1 = new byte[] { 0, 0, 0, 0 };
+
+            using (var pair = CreateNamedPipePair())
+            {
+                pair.Connect();
+                var server = pair.serverStream;
+                if (pair.writeToServer)
+                {
+                    using (NamedPipeClientStream client = new NamedPipeClientStream(PipeDirection.In, false, true, pair.clientStream.SafePipeHandle))
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            Assert.Equal(1, client.NumberOfServerInstances);
+                        }
+                        Task<int> clientTask = client.ReadAsync(received1, 0, received1.Length);
+                        server.Write(msg1, 0, msg1.Length);
+                        int receivedLength = await clientTask;
+                        Assert.Equal(msg1.Length, receivedLength);
+                        Assert.Equal(msg1, received1);
+                    }
+                }
+                else
+                {
+                    using (NamedPipeClientStream client = new NamedPipeClientStream(PipeDirection.Out, false, true, pair.clientStream.SafePipeHandle))
+                    {
+                        Task clientTask = client.WriteAsync(msg1, 0, msg1.Length);
+                        int receivedLength = server.Read(received1, 0, msg1.Length);
+                        Assert.Equal(msg1.Length, receivedLength);
+                        Assert.Equal(msg1, received1);
+                        await clientTask;
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public void ConnectOnAlreadyConnectedClient_Throws_InvalidOperationException()
+        {
+            using (var pair = CreateNamedPipePair())
+            {
+                var server = pair.serverStream;
+                var client = pair.clientStream;
+                byte[] buffer = new byte[] { 0, 0, 0, 0 };
+
+                pair.Connect();
+
+                Assert.True(client.IsConnected);
+                Assert.True(server.IsConnected);
+
+                Assert.Throws<InvalidOperationException>(() => client.Connect());
+            }
+        }
+
+        [Fact]
+        public async Task CancelTokenOn_ServerWaitForConnectionAsync_Throws_OperationCanceledException()
+        {
+            using (var pair = CreateNamedPipePair())
+            {
+                var server = pair.serverStream;
+                var ctx = new CancellationTokenSource();
+
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) // [ActiveIssue(812, PlatformID.AnyUnix)] - cancellation token after the operation has been initiated
+                {
+                    Task serverWaitTimeout = server.WaitForConnectionAsync(ctx.Token);
+                    ctx.Cancel();
+                    await Assert.ThrowsAnyAsync<OperationCanceledException>(() => serverWaitTimeout);
```

Comment:
Not pushing one way or the other, but why do we care whether it's an OperationCanceledException or the more derived TaskCanceledException?